File: kernel.c
#include <wiringPi.h>
#include <stdlib.h>

#include <signal.h>
#include <unistd.h>

int Sched_AddT(void (*f)(void), int d, int p);

typedef struct
{
    /* period in ticks */
    int period;
    /* ticks until next activation */
    int delay;
    /* function pointer */
    void (*func)(void);
    /* activation counter */
    int exec;
} Sched_Task_t;

void t1()
{
    printf("hello world 1sec");
}

void t2()
{
    printf("hello world 5sec");
}

void t3()
{
    printf("hello world 10sec");
}

#define NT 20
Sched_Task_t Tasks[NT];
int cur_task = NT;

int Sched_Init(void)
{
    for (int x = 0; x < NT; x++)
        Tasks[x].func = 0;

    wiringPiSetupPhys();
    signal(SIGALRM, alarmWakeup);
    ualarm(0, 1);
}

int Sched_AddT(void (*f)(void), int d, int p)
{
    for (int x = 0; x < NT; x++)
        if (!Tasks[x].func)
        {
            Tasks[x].period = p;
            Tasks[x].delay = d;
            Tasks[x].exec = 0;
            Tasks[x].func = f;
            return x;
        }
    return -1;
}

void Sched_Schedule(void)
{
    for (int x = 0; x < NT; x++)
    {
        if (Tasks[x].func)
        {
            if (Tasks[x].delay)
            {
                Tasks[x].delay--;
            }
            else
            {
                /* Schedule Task */
                Tasks[x].exec++;
                Tasks[x].delay = Tasks[x].period - 1;
            }
        }
    }
}

void Sched_Dispatch(void)
{
    int prev_task = cur_task;
    for (int x = 0; x < cur_task; x++)
    {
        if ((Tasks[x].func) && (Tasks[x].exec))
        {
            Tasks[x].exec = 0;
            cur_task = x;
            // interrupts();
            Tasks[x].func();
            // noInterrupts();
            cur_task = prev_task;
            /* Delete task if one-shot */
            if (!Tasks[x].period)
                Tasks[x].func = 0;
        }
    }
}

void setup()
{
    // initialize digital pin LED_BUILTIN as an output.
    Sched_Init();

    /* add a bunch of one-shot tasks, that will temporarily take up space in the TCB array */
    /* This forces task t1 to have a lower priority, and leave empty TCB entries for the
     *  toggle task added by t0.
     */
    Sched_AddT(t1, 0 /* delay */, 1000 /* period */);
    Sched_AddT(t2, 0 /* delay */, 5000 /* period */);
    Sched_AddT(t3, 0 /* delay */, 10000 /* period */);
}

void int_handler(int sig_num)
{
    if (sig_num != SIGALRM)
        return;
    Sched_Schedule();
    Sched_Dispatch();
}

int main()
{
    setup();
    while (1)
    {
        // nothing
    }
}

File: kernel.py
from task import Task
import threading

timer = None
tasks = []
current_task = 0

def t1():
    print("hello world 1sec")
    return

def t2():
    print("hello world 5secs")
    return

def t3():
    print("hello world 10secs")
    return

def set_interrupts():
    global timer
    timer = threading.Timer(0.1, Sched_Interrupt)
    timer.start()

def stop_interrupts():
    global timer
    timer.cancel()

def Sched_Init():
    # TODO configure interrupts and stuff on PI
    set_interrupts()
    return

def Sched_AddTask(func, delay, period):
    global current_task
    tasks.append(Task(func, delay, period))
    current_task = len(tasks)

def Sched_Schedule():
    for task in tasks:
        if task.func == None:
            continue
        if task.delay > 0:
            task.delay -= 1
        else:
            task.counter += 1
            task.delay = task.period - 1

def Sched_Dispatch():
    global current_task
    prev_task = current_task
    for i in range(current_task):
        task = tasks[i]
        if task.func == None or task.counter == 0:
            continue
        task.counter = 0
        current_task = i

        # enable interrupts
        set_interrupts()
        task.func()
        # disable interrupts
        stop_interrupts()

        current_task = prev_task

        # one shot task
        if task.period == 0:
            tasks.remove(task)

def setup():
    print("setting up")
    # TODO initialize pins
    Sched_Init()

    # TODO this is using seconds but needs to use ms later
    Sched_AddTask(t1, 0, 10)
    Sched_AddTask(t2, 0, 50)
    Sched_AddTask(t3, 0, 100)

def Sched_Interrupt():
    stop_interrupts()
    Sched_Schedule()
    Sched_Dispatch()
    set_interrupts()

setup()

while True:
    pass # keep it running
